src/libraries/System.Console/src/System/Console.cs

```diff
@@ -33,31 +34,54 @@ public static class Console
         private static ConsoleCancelEventHandler? s_cancelCallbacks;
         private static ConsolePal.ControlCHandlerRegistrar? s_registrar;
 
-        internal static T EnsureInitialized<T>([NotNull] ref T? field, Func<T> initializer) where T : class =>
-            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, initializer);
+        public static TextReader In
+        {
+            get
+            {
+                return Volatile.Read(ref s_in) ?? EnsureInitialized();
 
-        public static TextReader In => EnsureInitialized(ref s_in, () => ConsolePal.GetOrCreateReader());
+                static TextReader EnsureInitialized()
+                {
+                    // Must be placed outside s_syncObject lock. See Out getter.
+                    ConsolePal.EnsureConsoleInitialized();
+
+                    lock (s_syncObject) // Ensures In and InputEncoding are synchronized.
+                    {
+                        return s_in ??= ConsolePal.GetOrCreateReader();
+                    }
+                }
+            }
+        }
 
         public static Encoding InputEncoding
         {
             get
             {
-                return EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
+                Encoding? encoding = Volatile.Read(ref s_inputEncoding);
+                if (encoding == null)
+                {
+                    lock (s_syncObject)
+                    {
+                        encoding = s_inputEncoding ??= ConsolePal.InputEncoding;
+                    }
+                }
+                return encoding;
             }
             set
             {
                 CheckNonNull(value, nameof(value));
-                lock (InternalSyncObject)
+
+                lock (s_syncObject)
                 {
                     // Set the terminal console encoding.
                     ConsolePal.SetConsoleInputEncoding(value);
 
-                    Volatile.Write(ref s_inputEncoding, (Encoding)value.Clone());
+                    s_inputEncoding = (Encoding)value.Clone();
 
                     // We need to reinitialize 'Console.In' in the next call to s_in
                     // This will discard the current StreamReader, potentially
                     // losing buffered data.
-                    Volatile.Write(ref s_in, null);
+                    s_in = null;
```

Comment:
I realize this is the way it was before, but seems like it'd be slightly better to do this as the first thing in the lock?  That way a concurrent access to In will be forced to pick up the new encoding.  It's not a big deal, though; this was already like this, and with such concurrent access there are inherently races we can't do anything about, such that someone accessing both In and InputEncoding could always see inconsistent results due to one of them being changed in the interim.

Reply:
When I attempted to write this lock free, I had reversed the order.
But I left it as is, since it is inherently racy.

Reply:
That's fine.